op_crates/fetch/lib.rs

```diff
@@ -133,28 +141,79 @@ where
     return Err(type_error(format!("scheme '{}' not supported", scheme)));
   }
 
-  {
-    let state_ = state.borrow();
-    let permissions = state_.borrow::<FP>();
-    permissions.check_net_url(&url_)?;
-  }
+  let permissions = state.borrow::<FP>();
+  permissions.check_net_url(&url_)?;
 
   let mut request = client.request(method, url_);
 
-  match data.len() {
-    0 => {}
-    1 => request = request.body(Vec::from(&*data[0])),
-    _ => panic!("Invalid number of arguments"),
-  }
+  let maybe_request_body_rid = if args.has_body {
+    match data.len() {
+      0 => {
+        // If no body is passed, we return a writer for streaming the body.
+        let (tx, rx) = mpsc::channel::<Result<Vec<u8>, std::io::Error>>(1);
```

Comment:
```suggestion
        let (tx, rx) = mpsc::channel::<std::io::Result<Vec<u8>>>(1);
```
Seems to be the predominant pattern in cli/ and runtime/.